core/src/test/java/io/substrait/type/proto/JoinRoundtripTest.java

```diff
@@ -31,4 +32,18 @@ void hashJoin() {
             .build();
     verifyRoundTrip(relWithoutKeys);
   }
+
+  @Test
+  void nestedLoopJoin() {
+    Rel rel =
+        NestedLoopJoin.builder()
+            .from(
+                b.nestedLoopJoin(
+                    __ -> b.equal(b.fieldReference(leftTable, 0), b.fieldReference(rightTable, 2)),
```

Comment:
Ah! This is subtle and weird (and something to improve in the builder actually).

The condition on the join is evaluated relative to the joins record shape, which is the union of the left and right table records. It would look like:
```
(a, b, c, d, e, f)
```

`b.fieldReference(rightTable, 2)` here ends up referring to `c`

Reply:
Ah that makes sense! Thank you

Reply:
I'm realising there's nothing else in the code that uses this, which is why it hasn't come up before.

Reply:
For the test case, should I create a unioned named table to refer to? `5` is our of bounds when creating the Rel, but in bounds after the roundtrip

Reply:
Hmm, we have `newInputRelReference(int index, List<Rel> rels)`, which might handle this. I'll dig into this more myself.